src/main/java/org/tikv/common/region/AbstractRegionStoreClient.java

```diff
@@ -110,215 +108,200 @@ public boolean onNotLeader(TiRegion newRegion) {
-          updateClientStub();
-          return true;
-        } else {
-          continue;
-        }
-      } else if (candidateLeader.getId() == cur.getId()) {
-        hasVisitedStore = true;
-      }
-    }
-    candidateLeader = null;
-    retryLeaderTimes = peers.size();
-    return false;
+    return region.getReplicaContext(resolvedLocks, storeType);
   }
 
   private void updateClientStub() {
-    String addressStr = targetStore.getStore().getAddress();
-    if (targetStore.getProxyStore() != null) {
-      addressStr = targetStore.getProxyStore().getAddress();
+    String addressStr = store.getStore().getAddress();
+    if (store.getProxyStore() != null) {
+      addressStr = store.getProxyStore().getAddress();
     }
     ManagedChannel channel =
         channelFactory.getChannel(addressStr, regionManager.getPDClient().getHostMapping());
     blockingStub = TikvGrpc.newBlockingStub(channel);
-    asyncStub = TikvGrpc.newStub(channel);
-    if (targetStore.getProxyStore() != null) {
+    asyncStub = TikvGrpc.newFutureStub(channel);
+    if (store.getProxyStore() != null) {
       Metadata header = new Metadata();
-      header.put(TiConfiguration.FORWARD_META_DATA_KEY, targetStore.getStore().getAddress());
+      header.put(TiConfiguration.FORWARD_META_DATA_KEY, store.getStore().getAddress());
       blockingStub = MetadataUtils.attachHeaders(blockingStub, header);
       asyncStub = MetadataUtils.attachHeaders(asyncStub, header);
     }
   }
 
-  private boolean retryOtherStoreByProxyForward() {
-    if (!conf.getEnableGrpcForward()) {
-      return false;
+  private Metapb.Peer switchLeader() {
+    List<SwitchLeaderTask> responses = new LinkedList<>();
+    for (Metapb.Peer peer : region.getFollowerList()) {
+      ByteString key = region.getStartKey();
+      TiStore store = regionManager.getStoreById(peer.getStoreId());
+      ManagedChannel channel =
+          channelFactory.getChannel(
+              store.getAddress(), regionManager.getPDClient().getHostMapping());
+      TikvGrpc.TikvFutureStub stub =
+          TikvGrpc.newFutureStub(channel).withDeadlineAfter(timeout, TimeUnit.MILLISECONDS);
+      Kvrpcpb.RawGetRequest rawGetRequest =
+          Kvrpcpb.RawGetRequest.newBuilder()
+              .setContext(region.getReplicaContext(peer))
+              .setKey(key)
+              .build();
+      ListenableFuture<Kvrpcpb.RawGetResponse> task = stub.rawGet(rawGetRequest);
+      responses.add(new SwitchLeaderTask(task, peer));
     }
-    if (retryForwardTimes >= region.getFollowerList().size()) {
-      // If we try to forward request to leader by follower failed, it means that the store of old
-      // leader may be
-      // unavailable but the new leader has not been report to PD. So we can ban this store for a
-      // short time to
-      // avoid too many request try forward rather than try other peer.
-      if (originStore != null) {
-        originStore.forwardFail();
+    while (true) {
+      try {
```

Comment:
could you add these metrics for each request:
- the total backoff time
	- the total backoff time caused by grpc forward
	- the total backoff time caused by switch leader
- the total backoff operation
	- the total backoff operation for grpc forward
	- the total backoff operation for switch leader

with these metrics, we can see where the time is spent, and the detailed reason.